### PR TITLE
Avatar home page + library

### PR DESCRIPTION
**Pull request by kev306**

### CODE REVIEW

File: webpack.common.js [hunk 1]
```diff
@@ -2,6 +2,7 @@ const path = require('path');
 
 module.exports = {
   entry: {
+    avatarHome: './src/views/components/avatar/hydrate.tsx',
```
Comment (vck3000) [inline webpack.common.js:5]:
```suggestion
    avatarHome: './src/views/components/avatarHome/hydrate.tsx',
```

File: src/views/components/avatar/index.tsx [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
+import { hot } from 'react-hot-loader/root';
+import { AvatarHomeUi } from './avatarHome';
+
+function AvatarHome() {
+  return <AvatarHomeUi />;
```
Comment (vck3000) [inline src/views/components/avatar/index.tsx:6]:
```suggestion
  return <AvatarHome />;
```

File: src/views/components/avatar/avatarHome.tsx [hunk 3]
```diff
@@ -0,0 +1,182 @@
+import React, { useEffect, useState } from 'react';
+import Swal from 'sweetalert2';
+
+export function AvatarHomeUi() {
+  const [currentResImgLink, setCurrentResImgLink] = useState(
+    '../../avatars/base-res.svg',
```
Comment (vck3000) [inline src/views/components/avatar/avatarHome.tsx:6]:
```suggestion
    '/assets/avatars/base-res.svg',
```

File: src/views/components/avatar/avatarHome.tsx [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+  );
+  const [currentSpyImgLink, setCurrentSpyImgLink] = useState(
+    '../../avatars/base-spy.svg',
+  );
+  const [avatarLibrary, setAvatarLibrary] = useState([]);
```
Comment (vck3000) [inline src/views/components/avatar/avatarHome.tsx:11]: Can you type this guy?

File: src/views/components/avatar/avatarHome.tsx [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+  const [selectedAvatarId, setSelectedAvatarId] = useState(null);
+  const [selectedAvatarResLink, setSelectedAvatarResLink] = useState(null);
+  const [selectedAvatarSpyLink, setSelectedAvatarSpyLink] = useState(null);
+
+  useEffect(() => {
+    // TODO-kev: Figure out why this runs 8 times
+    async function fetchUserAvatarInfo() {
+      // TODO-kev: Remove hardcoded username
+      const response = await fetch('/profile/1/avatar/avatarinfo');
+      const data = await response.json();
```
Comment (vck3000) [inline src/views/components/avatar/avatarHome.tsx:21]: Let's type this data guy too

File: src/views/components/avatar/avatarHome.tsx [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+
+      // TODO-kev: Remove the loading for this?
```
Comment (vck3000) [inline src/views/components/avatar/avatarHome.tsx:23]: Option 1: static scaffold image
Option 2: loading gif

File: src/views/components/avatar/avatarHome.tsx [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+      }
+      if (data.currentSpyLink) {
+        setCurrentSpyImgLink(data.currentSpyLink);
+      }
+
+      setAvatarLibrary(data.avatarLibrary);
+    }
+
+    // TODO-kev: What to add for .then() or .catch()? valid json issue
+    fetchUserAvatarInfo();
```
Comment (vck3000) [inline src/views/components/avatar/avatarHome.tsx:35]:
```suggestion
    void fetchUserAvatarInfo();
```

File: src/views/components/avatar/avatarHome.tsx [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+          },
+          body: JSON.stringify({
+            avatarId: selectedAvatarId,
+            resLink: selectedAvatarResLink,
+            spyLink: selectedAvatarSpyLink,
+          }),
+        });
+
+        if (response.status === 200) {
+          Swal.close();
```
Comment (vck3000) [inline src/views/components/avatar/avatarHome.tsx:64]: Let's move this down next to the Swal.fire()

File: src/routes/profile.tsx [hunk 9]
```diff
@@ -14,10 +14,17 @@ import { S3Agent } from '../clients/s3/S3Agent';
 import { PatreonAgent } from '../clients/patreon/patreonAgent';
```
Comment (vck3000) [inline src/routes/profile.tsx:14]: Let's make a new folder called `profile`, keep the profile.ts here, and then have it import `profile/profile.ts` and `profile/avatar.ts`

File: src/routes/profile.tsx [hunk 10]
```diff
@@ -171,9 +178,134 @@ router.post(
   },
 );
 
-// Show the customavatar edit page
+// Show the user's avatar homepage
+router.get(
+  '/:profileUsername/avatarhome',
+  checkProfileOwnership,
+  async (req, res) => {
```
Comment (vck3000) [inline src/routes/profile.tsx:185]: Let's try to type these guys.

File: src/routes/profile.tsx [hunk 11]
```diff
@@ -171,9 +178,134 @@ router.post(
...
+      approvedAvatarSets,
+      currentPatreon,
+      avatarHomeReact,
+    });
+  },
+);
+
+// Get a users avatar library links
+router.get(
+  '/:profileUsername/avatar/avatarinfo',
```
Comment (vck3000) [inline src/routes/profile.tsx:221]:
```suggestion
  '/:profileUsername/avatar/getAllAvatars',
```

File: src/routes/profile.tsx [hunk 12]
```diff
@@ -171,9 +178,134 @@ router.post(
...
+  async (req, res) => {
+    const user = await User.findOne({
+      usernameLower: req.user.usernameLower,
+    });
+
+    const result = {
+      currentResLink: user.avatarImgRes,
+      currentSpyLink: user.avatarImgSpy,
+      avatarLibrary: await s3Agent.getUsersAvatarLibraryLinks(
+        user.usernameLower,
```
Comment (vck3000) [inline src/routes/profile.tsx:232]: Let's change this to provide in the avatar library with avatar sets and "transform" them into links so the agent itself doesn't need to access the user model.

File: src/clients/s3/S3Agent.ts [hunk 13]
```diff
@@ -176,4 +184,63 @@ export class S3Agent {
...
+          `${FolderName.APPROVED}/${usernameLower}/${usernameLower}_spy_${id}.png`,
+        ),
+      };
+      avatarLibraryLinks.push(avatarSet);
+    });
+
+    return avatarLibraryLinks;
+  }
+
+  private async updateUsersAvatarLibrary(usernameLower: string) {
```
Comment (vck3000) [inline src/clients/s3/S3Agent.ts:212]: Make this guy be called by getRewards each time.

File: src/clients/s3/S3Agent.ts [hunk 14]
```diff
@@ -13,11 +13,19 @@ interface S3AvatarLinks {
...
+  resLink: string;
+  spyLink: string;
+}
+
 export interface IS3Controller {
   listObjectKeys(prefixes: string[]): Promise<string[]>;
   uploadFile(key: string, fileContent: Buffer, contentType: string): any;
   deleteFile(link: string): any;
   moveFile(oldLink: string, newLink: string): any;
+  getLinkFromKey(key: string): string;
```
Comment (vck3000) [inline src/clients/s3/S3Agent.ts:28]:
```suggestion
  transformKeyToLink(key: string): string;
```

File: src/clients/s3/S3Agent.ts [hunk 15]
```diff
@@ -13,11 +13,19 @@ interface S3AvatarLinks {
...
+interface ApprovedAvatarSet {
+  avatarSetId: number;
+  resLink: string;
+  spyLink: string;
+}
+
 export interface IS3Controller {
   listObjectKeys(prefixes: string[]): Promise<string[]>;
   uploadFile(key: string, fileContent: Buffer, contentType: string): any;
   deleteFile(link: string): any;
```
Comment (vck3000) [inline src/clients/s3/S3Agent.ts:26]: Update the return types!

File: src/rewards/getRewards.ts [hunk 16]
```diff
@@ -98,10 +98,14 @@ export async function userHasReward(
 export async function getAvatarLibrarySizeForUser(
   usernameLower: string,
 ): Promise<number> {
+  if (isMod(usernameLower)) {
+    return 10;
```
Comment (vck3000) [inline src/rewards/getRewards.ts:102]:
```suggestion
    return 5;
```

File: src/clients/s3/S3Controller.ts [hunk 17]
```diff
@@ -151,4 +151,8 @@ export default class S3Controller implements IS3Controller{
   private isValidLink(link: string) {
     return link.startsWith(this.publicFileLinkPrefix);
   }
+
+  public getLinkFromKey(key: string) {
```
Comment (vck3000) [inline src/clients/s3/S3Controller.ts:155]:
```suggestion
  public transformKeyToLink(key: string) {
```

File: src/clients/s3/S3Agent.ts [hunk 18]
```diff
@@ -13,11 +13,19 @@ interface S3AvatarLinks {
   spyLink: string;
 }
 
+// TODO-kev: Consider merging this with above
+interface ApprovedAvatarSet {
```
[the remark this answers is not in the document]
Comment (kev306) [reply src/clients/s3/S3Agent.ts]: Initial thoughts from reviewing: The S3AvatarLinks interface is integrated in multiple files that will take time to merge. Considering whether to do it in this PR or a separate one

[the remark this answers is not in the document]
Comment (kev306) [reply src/clients/s3/S3Agent.ts]: Change implemented within this pr on second thought

File: src/routes/profile/profile.tsx [hunk 19]
```diff
@@ -1,31 +1,21 @@
 // @ts-nocheck
```
Comment (vck3000) [inline src/routes/profile/profile.tsx:1]: Let's rename this to index.tsx.

File: src/clients/s3/S3Agent.ts [hunk 20]
```diff
@@ -1,31 +1,26 @@
...
-  deleteFile(link: string): any;
-  moveFile(oldLink: string, newLink: string): any;
-  getLinkFromKey(key: string): string;
+  uploadFile(
+    key: string,
+    fileContent: Buffer,
+    contentType: string,
+  ): Promise<string>;
+  deleteFile(link: string): void;
+  moveFile(oldLink: string, newLink: string): void;
```
Comment (vck3000) [inline src/clients/s3/S3Agent.ts:22]: Promise if it's async

File: src/rewards/getRewards.ts [hunk 21]
```diff
@@ -116,3 +122,40 @@ export async function getAvatarLibrarySizeForUser(
     return 10;
   }
 }
+
+async function updateUsersAvatarLibrary(
+  usernameLower: string,
+  patreonReward: any,
+) {
+  const user = await User.findOne({ usernameLower });
```
Comment (vck3000) [inline src/rewards/getRewards.ts:130]: Can you replace this with the user get adapter.

File: src/routes/profile/avatarRoutes.tsx [hunk 22]
```diff
@@ -0,0 +1,94 @@
+// @ts-nocheck
```
Comment (vck3000) [inline src/routes/profile/avatarRoutes.tsx:1]:
```suggestion

```

File: src/routes/profile/avatarRoutes.tsx [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+// TODO-kev: Add something similar to checkProfileOwnership? Do i even need this?
+
+// TODO-kev: Should i keep this here? Or move back to profile
+// Show the user's avatar homepage
+router.get('/', (req, res) => {
+  const avatarHomeReact = renderToString(<AvatarHome />);
+  res.render('profile/avatarhome', { avatarHomeReact });
+});
+
+// Change a users current avatar
```
Comment (vck3000) [inline src/routes/profile/avatarRoutes.tsx:38]:
```suggestion
// For a user to change their own current avatar
```

File: src/routes/profile/avatarRoutes.tsx [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  user.avatarImgRes = req.body.resLink;
+  user.avatarImgSpy = req.body.spyLink;
+  await user.save();
+
+  res.status(200).send('Successfully changed avatar.');
+});
+
+// Get a users avatar library links
```
Comment (vck3000) [inline src/routes/profile/avatarRoutes.tsx:73]:
```suggestion
// Get a user's avatar library links
```

File: src/routes/profile/avatarRoutes.tsx [hunk 25]
```diff
@@ -0,0 +1,94 @@
...
+router.get('/getallavatars', async (req, res) => {
```
Comment (vck3000) [inline src/routes/profile/avatarRoutes.tsx:74]:
```suggestion
router.get('/getalluseravatars', async (req, res) => {
```

File: src/routes/profile/avatarRoutes.tsx [hunk 26]
```diff
@@ -0,0 +1,94 @@
...
+  // TODO-kev: Put this function here or when the avatar homepage is first rendered?
+  await getAndUpdatePatreonRewardTierForUser(req.user.usernameLower);
+
+  const user = await User.findOne({
```
Comment (vck3000) [inline src/routes/profile/avatarRoutes.tsx:78]: database adatper

File: src/routes/profile/avatarRoutes.tsx [hunk 27]
```diff
@@ -0,0 +1,94 @@
...
+    usernameLower: req.user.usernameLower,
+  });
+
+  const result: AllAvatarsRouteReturnType = {
+    currentResLink: user.avatarImgRes,
+    currentSpyLink: user.avatarImgSpy,
+    avatarLibrary: await s3Agent.getUsersAvatarLibraryLinks(
+      user.usernameLower,
+      user.avatarLibrary as number[],
```
Comment (vck3000) [inline src/routes/profile/avatarRoutes.tsx:87]: Add this to IUser please to get rid of `as number[]`

File: src/sockets/commands/mod/mpushavatartolibrary.ts [hunk 28]
```diff
@@ -48,12 +47,9 @@ export const mpushavatartolibrary: Command = {
       return;
     }
 
-    const librarySize = await getAvatarLibrarySizeForUser(usernameLower);
```
Comment (vck3000) [inline src/sockets/commands/mod/mpushavatartolibrary.ts:51]: Let's revert this

File: src/views/components/avatarHome/avatarHome.tsx [hunk 29]
```diff
@@ -1,26 +1,49 @@
...
   useEffect(() => {
-    // TODO-kev: Figure out why this runs 8 times
+    require('./styles.css');
+    require('react-multi-carousel/lib/styles.css');
+
     async function fetchUserAvatarInfo() {
-      // TODO-kev: Remove hardcoded username
-      const response = await fetch('/profile/1/avatar/avatarinfo');
-      const data = await response.json();
+      const response = await fetch('/profile/avatar/getallavatars');
```
Comment (vck3000) [inline src/views/components/avatarHome/avatarHome.tsx:44]: Can you pull out these routes to a const at the top

File: src/views/components/avatarHome/avatarHome.tsx [hunk 30]
```diff
@@ -83,6 +107,27 @@ export function AvatarHomeUi() {
...
+    setSelectedAvatarId(avatarSet.avatarSetId);
+    setSelectedAvatarResLink(avatarSet.resLink);
+    setSelectedAvatarSpyLink(avatarSet.spyLink);
+  };
+
+  if (isLoading) {
+    return (
+      <div className="loading-container">
+        <iframe
+          src="https://giphy.com/embed/swhRkVYLJDrCE"
```
Comment (vck3000) [inline src/views/components/avatarHome/avatarHome.tsx:120]: Replace with css

**Review by vck3000 — CHANGES_REQUESTED** [review]
-

File: src/rewards/getRewards.ts [hunk 31]
```diff
@@ -102,33 +109,42 @@ export async function userHasReward(
   return true;
 }
 
-function getAvatarLibrarySizeForUser(
+export async function getAvatarLibrarySizeForUser(
```
Comment (vck3000) [inline src/rewards/getRewards.ts:112]:
```
export async function getAvatarLibrarySizeForUser(
  usernameLower: string,
  patreonReward?: RewardType,
): Promise<number> {

  const patreonLibrarySize = async () => {
    if (!patreonReward) {
      patreonReward = await getPatreonRewardTierForUser(usernameLower);
    }

    if (patreonReward === constants.TIER1_BADGE) {
      return 2;
    } else if (patreonReward === constants.TIER2_BADGE) {
      return 3;
    } else if (patreonReward === constants.TIER3_BADGE) {
      return 5;
    } else if (patreonReward === constants.TIER4_BADGE) {
      return 10;
    } else {
      return 0;
    }
  };

 const modLibrarySize = () => {
   return isMod(usernameLower) ? 5 : 0;
  };

  return Math.max(await patreonLibrarySize(), modLibrarySize());
}
```

File: src/views/components/avatarHome/avatarHome.tsx [hunk 32]
```diff
@@ -5,20 +5,31 @@ import { S3AvatarSet } from '../../../clients/s3/S3Agent';
...
+    breakpoint: { max: 732, min: 464 },
+    items: 1,
+  },
   mobile: {
     breakpoint: { max: 464, min: 0 },
     items: 1,
   },
 };
 
+type AvatarHomeRoutes = {
```
Comment (vck3000) [inline src/views/components/avatarHome/avatarHome.tsx:26]: Let's make this an interface

File: src/views/components/avatarHome/avatarHome.tsx [hunk 33]
```diff
@@ -5,20 +5,31 @@ import { S3AvatarSet } from '../../../clients/s3/S3Agent';
...
+  getalluseravatars: string;
+  changeavatar: string;
+  customavatar: string;
+  edit: string;
```
Comment (vck3000) [inline src/views/components/avatarHome/avatarHome.tsx:30]:
```
const GetLinks = {
  getalluseravatars: (username: string) => `/profile/${username}/avatar/getalluseravatars`,
}
```

File: src/databaseAdapters/user.ts [hunk 34]
```diff
@@ -12,3 +12,6 @@ export class MongoUserAdapter implements DatabaseAdapter {
     })) as IUser;
   }
 }
+
+const userAdapter = new MongoUserAdapter();
```
Comment (vck3000) [inline src/databaseAdapters/user.ts:16]: Let's remove the export on L8

File: src/gameplay/types.ts [hunk 35]
```diff
@@ -71,6 +71,9 @@ export interface IUser {
   IPAddresses?: string[];
   lastIPAddress?: string;
   matchmakingBlacklist?: string[];
+
+  save: () => Promise<void>;
```
Comment (vck3000) [inline src/gameplay/types.ts:75]: Can u add a comment above saying

```
// Mongoose methods
```

File: src/rewards/getRewards.ts [hunk 36]
```diff
@@ -113,34 +113,43 @@ export async function getAvatarLibrarySizeForUser(
...
+    if (patreonReward === constants.TIER1_BADGE) {
+      return 2;
+    } else if (patreonReward === constants.TIER2_BADGE) {
+      return 3;
+    } else if (patreonReward === constants.TIER3_BADGE) {
+      return 5;
+    } else if (patreonReward === constants.TIER4_BADGE) {
+      return 10;
+    } else {
+      return 0;
```
Comment (vck3000) [inline src/rewards/getRewards.ts:137]:
```suggestion
      return TEMP_MIN_LIBRARY_SIZE; // 1;
```

File: src/rewards/getRewards.ts [hunk 37]
```diff
@@ -113,34 +113,43 @@ export async function getAvatarLibrarySizeForUser(
   usernameLower: string,
   patreonReward?: RewardType,
 ): Promise<number> {
-  let librarySize = 0;
+  // Temporary minLibrarySize for feature launch. To be disabled DD/MM/YYYY
+  const tempMinLibrarySize = 2;
```
Comment (vck3000) [inline src/rewards/getRewards.ts:117]: Can we pull this const to the top.

File: src/rewards/getRewards.ts [hunk 38]
```diff
@@ -113,34 +113,43 @@ export async function getAvatarLibrarySizeForUser(
...
+    } else {
+      return 0;
+    }
+  };
 
-  return librarySize;
+  return Math.max(
+    await patreonLibrarySize(),
+    modLibrarySize(),
+    tempMinLibrarySize,
```
Comment (vck3000) [inline src/rewards/getRewards.ts:144]:
```suggestion
    TEMP_MIN_LIBRARY_SIZE,
```

File: src/views/components/avatarHome/avatarHome.tsx [hunk 39]
```diff
@@ -23,65 +23,61 @@ const responsive = {
...
-
-    const newRoutes: AvatarHomeRoutes = {
-      getalluseravatars: `/profile/${extractedUsername}/avatar/getalluseravatars`,
-      changeavatar: `/profile/${extractedUsername}/avatar/changeavatar`,
-      customavatar: `/profile/${extractedUsername}/customavatar`,
-      edit: `/profile/${extractedUsername}/edit`,
-    };
-
-    setRoutes(newRoutes);
+    setUsername(extractedUsername);
```
Comment (vck3000) [inline src/views/components/avatarHome/avatarHome.tsx:56]: Try to move the 2nd useEffect stuff here and use `extractedUsername`

File: src/views/components/avatarHome/avatarHome.tsx [hunk 40]
```diff
@@ -23,65 +23,61 @@ const responsive = {
...
       const data: AllAvatarsRouteReturnType = await response.json();
 
-      if (data.currentResLink) {
-        setCurrentResImgLink(data.currentResLink);
-      }
-      if (data.currentSpyLink) {
-        setCurrentSpyImgLink(data.currentSpyLink);
-      }
+      setCurrentResImgLink(
+        data.currentResLink ? data.currentResLink : '/avatars/base-res.svg',
```
Comment (vck3000) [inline src/views/components/avatarHome/avatarHome.tsx:71]: Let's move `/avatars/base-res.svg` to a const at the top or in a higher level file.

File: src/views/components/avatarHome/avatarHome.tsx [hunk 41]
```diff
@@ -124,14 +120,7 @@ export function AvatarHome() {
...
       <div className="loading-container">
-        <iframe
-          src="https://giphy.com/embed/swhRkVYLJDrCE"
-          width="480"
-          height="270"
-          frameBorder="0"
-          className="giphy-embed"
-          allowFullScreen
-        ></iframe>
+        <div className="loader"></div>
```
Comment (vck3000) [inline src/views/components/avatarHome/avatarHome.tsx:123]: Let's give this a more specific class name.

File: src/views/components/avatarHome/avatarHome.tsx [hunk 42]
```diff
@@ -159,7 +148,7 @@ export function AvatarHome() {
           src={currentSpyImgLink}
         ></img>
       </div>
-      <a className="btn btn-info" href={routes.customavatar}>
+      <a className="btn btn-info" href={GetLinks.customavatar(username)}>
```
Comment (vck3000) [inline src/views/components/avatarHome/avatarHome.tsx:151]:
```suggestion
      <a className="btn btn-info" href={getLinks.customavatar(username)}>
```

File: src/views/components/avatarHome/avatarHome.tsx [hunk 43]
```diff
@@ -228,11 +217,17 @@ export function AvatarHome() {
         </h4>
       </div>
 
+      <h4>
+        <strong style={{ color: '#1976D2' }}>
+          Until the DD/MM/YYYY, all users will be given a minimum avatar library
+          size of 2!
+        </strong>
+      </h4>
```
Comment (vck3000) [inline src/views/components/avatarHome/avatarHome.tsx:225]: Let's add the patreon stuff here in like 4 dot points or something super simple.

File: src/views/components/avatarHome/avatarHome.tsx [hunk 44]
```diff
@@ -228,11 +217,17 @@ export function AvatarHome() {
         </h4>
       </div>
 
+      <h4>
+        <strong style={{ color: '#1976D2' }}>
+          Until the DD/MM/YYYY, all users will be given a minimum avatar library
```
Comment (vck3000) [inline src/views/components/avatarHome/avatarHome.tsx:222]: For a limited time until <date>, all users ...

File: src/gameplay/types.ts [hunk 45]
```diff
@@ -43,7 +43,7 @@ export interface IUser {
   emailTokenExpiry?: Date;
```
Comment (vck3000) [inline src/gameplay/types.ts:43]: Let's add that automated check to limit number of avatar requests if they've hit their max limit of library stuff.

File: src/rewards/getRewards.ts [hunk 46]
```diff
@@ -134,14 +133,15 @@ export async function getAvatarLibrarySizeForUser(
     } else if (patreonReward === constants.TIER4_BADGE) {
       return 10;
     } else {
-      return 0;
+      return 1;
     }
   };
 
+  // TODO-kev: Temporary minLibrarySize for feature launch. To be disabled DD/MM/YYYY
```
Comment (vck3000) [inline src/rewards/getRewards.ts:140]: Remember to update before merging

File: src/routes/profile/index.tsx [hunk 47]
```diff
@@ -326,13 +329,22 @@ async function validateUploadAvatarRequest(
     }
   }
 
-  // Check: Both a singular res and spy avatar were submitted
+  // Check: Does not exceed avatar library size
+  const librarySize = await getAvatarLibrarySizeForUser(user.usernameLower);
+  if (user.avatarLibrary && user.avatarLibrary.length >= librarySize) {
+    return {
+      valid: false,
+      errMsg: `You have exceeded your maximum number of avatars: ${librarySize}. Please support our Patreon to increase this limit.`,
```
Comment (vck3000) [inline src/routes/profile/index.tsx:337]:
```suggestion
      errMsg: `You have exceeded your maximum number of avatars: ${librarySize}.`,
```

File: src/routes/profile/index.tsx [hunk 48]
```diff
@@ -326,13 +329,22 @@ async function validateUploadAvatarRequest(
     }
   }
 
-  // Check: Both a singular res and spy avatar were submitted
+  // Check: Does not exceed avatar library size
```
Comment (vck3000) [inline src/routes/profile/index.tsx:332]:
```suggestion
  // Check: User has space in their user library
```

File: src/routes/profile/index.tsx [hunk 49]
```diff
@@ -326,13 +329,22 @@ async function validateUploadAvatarRequest(
...
+    };
+  }
+
   if (
     !files['avatarRes'] ||
     !files['avatarSpy'] ||
     files['avatarRes'].length !== 1 ||
     files['avatarSpy'].length !== 1
   ) {
+    // Check: Both a singular res and spy avatar were submitted
```
Comment (vck3000) [inline src/routes/profile/index.tsx:347]: Can you move this comment above the if

File: src/views/components/avatarHome/avatarHome.tsx [hunk 50]
```diff
@@ -50,34 +52,32 @@ export function AvatarHome() {
...
     setUsername(extractedUsername);
-  }, []);
-
-  useEffect(() => {
-    setIsLoading(true);
 
-    if (username) {
-      void fetchUserAvatarInfo();
-    }
+    void fetchUserAvatarInfo();
```
Comment (vck3000) [inline src/views/components/avatarHome/avatarHome.tsx:62]: Let's have this call below fetchUserAvatarInfo()

File: src/routes/profile/index.tsx [hunk 51]
```diff
@@ -492,6 +486,11 @@ router.get(
   },
 );
 
+// Temporary redirect to the profile edit page. Used where the profileUsername is not easily obtained
+router.get('/edit/redirect', async (req, res) => {
+  return res.redirect(`/profile/${req.user.username}/edit`);
+});
```
Comment (vck3000) [inline src/routes/profile/index.tsx:492]: rm

File: src/sockets/commands/mod/mpushavatartolibrary.ts [hunk 52]
```diff
@@ -50,11 +50,11 @@ export const mpushavatartolibrary: Command = {
 
     const librarySize = await getAvatarLibrarySizeForUser(usernameLower);
     user.avatarLibrary.push(toBeAddedAvatarId);
+    user.markModified('avatarLibrary');
```
Comment (vck3000) [inline src/sockets/commands/mod/mpushavatartolibrary.ts:53]: This should be below L56

**Review by vck3000 — APPROVED** [review]

**Pull request merged**